### DISCUSSION TITLE
Rspec for batches complete callback

### DISCUSSION THREAD

**Discussion by haydengomes** [sidekiq/sidekiq#6483] [Q&A]
Hi, I'm trying to write tests for an `on_complete` callback. In the test suite, I've tried to run things with `Sidekiq::Testing.inline! { described_class.perform_async }`. However, if I put a breakpoint in the `on_complete` handler, I find that it shows a total of 4, with 0 failures, 0 completed, and 1 pending. If I call `status.complete?` it returns false.

Am I misunderstanding how inline works that it seems to be triggering the complete handler without having actually run any of the jobs? I thought from the description for the callback type that `completed` should be the same as `total`

**Comment by kevinrobell-st** [comment]
You need to add the batches middleware before `Sidekiq::Testing.inline!` will start working.
- https://github.com/sidekiq/sidekiq/wiki/Batches#testing

Note: This doesn't preserve the expected order of nested batches
- https://github.com/sidekiq/sidekiq/issues/6330

**Reply by haydengomes** [reply]
So we have added the middleware and a breakpoint in our `on_complete` method does get hit. So I think the inline is working in that I imagine the on_complete wouldn't get called before the test finished otherwise.

For more context, the worker is effectively:
```
class BatchingWorker < ApplicationWorker

def perform
    start_time = Time.now
    batch = Sidekiq::Batch.new
    batch.on(:complete, BatchingWorker, {'start_time' => start_time} )

    batch.jobs do
      ChildWorker.perform_bulk(Resource.all.map{|resource| [resource.id]})
    end
end

def on_complete(status, options)
    end_time = Time.now
    elapsed_time = end_time - Time.parse(options['start_time'])
    byebug
...
end
end
```

So when that byebug line is reached, the status looks like the following:
```
<Sidekiq::Batch::Status:######### @bid="########", @props={"callbacks"=>"{\"complete\":[{\"ChildWorker\":{\"start_time\":\"2024-10-30 20:54:39 +0000\"}}]}", "cbq"=>"queue_name", "created_at"=>"1730321679.7028239", "pending"=>"1", "kids"=>"0", "description"=>"#########", "total"=>"4"}, @failures=0, @completed=0, @pending=1, @total=4, @callbacks={"complete"=>[{"BatchingWorker"=>{"start_time"=>"2024-10-30 20:54:39 +0000"}}]}>
```

Also as a side question, I only have 3 Resource records, yet the batch shows a total of 4. does it include BatchingWorker in that count?
